src/wp-includes/interactivity-api/interactivity-api.php

```diff
@@ -0,0 +1,136 @@
+ *
+ * This processes only one root interactive block at a time because the
+ * rendered HTML of that block contains the rendered HTML of all its inner
+ * blocks, including any interactive block. It does so by ignoring all the
+ * interactive inner blocks until the root interactive block is processed.
+ *
+ * @since 6.5.0
+ *
+ * @param array $parsed_block The parsed block.
+ * @return array The same parsed block.
+ */
+function wp_interactivity_process_directives_of_interactive_blocks( $parsed_block ) {
+	static $root_interactive_block = null;
+
+	/*
+	 * Checks whether a root interactive block is already annotated for
+	 * processing, and if it is, it ignores the subsequent ones.
+	 */
+	if ( null === $root_interactive_block ) {
+		$block_name = $parsed_block['blockName'];
+		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
+
+		if ( isset( $block_name ) && isset( $block_type->supports['interactivity'] ) && $block_type->supports['interactivity'] ) {
+			// Annotates the root interactive block for processing.
+			$root_interactive_block = array( $block_name, md5( serialize( $parsed_block ) ) );
+
+			/*
+			 * Adds a filter to process the root interactive block once it has
+			 * finished rendering.
+			 */
+			$process_interactive_blocks = static function ( $content, $parsed_block ) use ( &$root_interactive_block, &$process_interactive_blocks ) {
+				// Checks whether the current block is the root interactive block.
+				list($root_block_name, $root_block_md5) = $root_interactive_block;
+				if ( $root_block_name === $parsed_block['blockName'] && md5( serialize( $parsed_block ) ) === $root_block_md5 ) {
+					// The root interactive blocks has finished rendering, process it.
+					$content = wp_interactivity_process_directives( $content );
+					// Removes the filter and reset the root interactive block.
+					remove_filter( 'render_block', $process_interactive_blocks );
+					$root_interactive_block = null;
+				}
+				return $content;
+			};
+
+			/*
+			 * Uses a priority of 20 to ensure that other filters can add additional
+			 * directives before the processing starts.
+			 */
+			add_filter( 'render_block', $process_interactive_blocks, 20, 2 );
+		}
+	}
+
+	return $parsed_block;
+}
+add_filter( 'render_block_data', 'wp_interactivity_process_directives_of_interactive_blocks', 10, 1 );
+
+/**
+ * Retrieves the main WP_Interactivity_API instance.
+ *
+ * It provides access to the WP_Interactivity_API instance, creating one if it
+ * doesn't exist yet. It also registers the hooks and necessary script
+ * modules.
+ *
+ * @since 6.5.0
+ *
+ * @return WP_Interactivity_API The main WP_Interactivity_API instance.
+ */
+function wp_interactivity() {
+	static $instance = null;
+	if ( is_null( $instance ) ) {
+		$instance = new WP_Interactivity_API();
+		$instance->add_hooks();
+		$instance->register_script_modules();
+	}
+	return $instance;
```

Comment:
For consistency with https://github.com/WordPress/wordpress-develop/pull/5945:

```suggestion
 *
 * @global WP_Interactivity_API $wp_interactivity
 *
 * @since 6.5.0
 *
 * @return WP_Interactivity_API The main WP_Interactivity_API instance.
 */
function wp_interactivity(): WP_Interactivity_API {
	global $wp_interactivity;

	if ( ! ( $wp_interactivity instanceof WP_Interactivity_API ) ) {
		$wp_interactivity = new WP_Interactivity_API();
	}

	return $wp_interactivity;
```

And then the calls to `add_hooks()` and `register_script_modules()` could be added to `wp-settings.php` in the same way.

Reply:
Done in https://github.com/WordPress/wordpress-develop/pull/5953/commits/69c6f78bd2a26246d0980497931042b43f8cad85.